### DISCUSSION TITLE
Another AVX optimisation proposal

### DISCUSSION THREAD

**Discussion by daliborh** [animetosho/md5-optimisation#7] [Q&A]
Hello,

In each step of MD5, we have to add constant to hashed value. But we can do it with AVX instructions before we begin computation in steps. It can be simply done by precomputing all 64 constants before hashing or 'per partes' like this:

- add first 16 constants (512bit) to the [message] (512bit)
- perform ROUND_F without adding constants, but only that modified message parts
- add modified 16 constants to the [message], so in [message] we will now have sums of original message and constants for G round. It means that these modified constatns are just differences between G constants and F constants...
- perform ROUND_G without adding constants, but only that modified message
- add modified 16 constants to the [message]...
- perferm ROUND_H...
- etc... 

With AVX-512, adittions for each round can be done in just 1 VPADDD instruction. With AVX2 in 2 instructions. Of course, some MOVs is also needed.
Do you think it can be an improvement?

**Comment by animetosho** [comment]
Thanks for the idea - it can work, but the main issue is with extracting all the lanes after the add (the "some MOVs is also needed" in your post).

If you stick to 128-bit vectors, this can be done via `[V]PEXTRD`, however if you go wider, you'll also need to extract the relevant 128-bit lane before extracting the 32-bit component. These are all considered 'shuffle' operations, and `PEXTRD` is also a cross domain move (FP -> integer), so it's not exactly a cheap operation. I recall trying it on Silvermont and getting a slight performance boost, so it seems to have merit on that specific processor, though it's unlikely to be beneficial in the general case.

The idea is more sensible if you're already using SIMD for everything else - as such, the AVX-512 version [uses the idea](https://github.com/animetosho/md5-optimisation/blob/31fc4e985ca9efbed5810c665d415b19e089ebee/md5-x86-asm.h#L1241-L1243), though I didn't write about it in the article. The implementation does also stick to 128-bit vectors (avoids cross-lane operations as well as frequency throttling on Intel CPUs, or the "512-bit mode penalty" disabling EUs).

Also keep in mind that adding constants isn't part of the critical path, so for most cases, it won't show much/any improvement even if it is better. It *could* be beneficial on 'narrow' processors (such as older Intel Atom), but it's not so straightforward (have to balance cost of extracting lanes).

Your idea about using the difference between F/G constants is something I hadn't considered before though, and a neat trick to save some registers. I doubt it yields any practical benefit, but it doesn't seem to have any downside, and is a neat trick nonetheless!

**Reply by daliborh** [reply]
Thank you for reply.

I thought that after VPADDD, you do not have to shuffle the vector to extract 32 bit values, you can just use one VMOVDQA (?) instruction to save the whole 512 bits back into memory and then in each step just get it from there. It should fit into cache, so the latency connected with writing zmm or ymm register to memory maybe won't hurt it much, considering that those MOVs can be done in parallel with other operations a lot instructions before values are used...

Simply said - you have to read constants from the memory somehow anyway, so if they are saved there beforehand, maybe it won't hurt :-)

**Reply by animetosho** [reply]
With reading from recently written memory, you'd need to check the store-to-load forwarding behaviour on the processor. [This article](https://chipsandcheese.com/2022/11/08/amds-zen-4-part-2-memory-subsystem-and-conclusion/) shows a 6-7 cycle latency on Zen4 for loads that overlap the store, but this is measured using 64b stores and 32b loads. I don't know if the behaviour is different for vector stores with int loads, and you'd have to check it across a few processors.

If you tried to actually write this using intrinsics, you'd find the compiler just converts it back to using shuffle/extract, which makes me think that store-forwarding doesn't handle these cases well.

Even if it does work well, keep in mind that you still have to add each lane back into `A`, and that most implementations use `LEA` to do two adds in a single instruction, so it's questionable if it even reduces the instruction count.

But if you're willing to try it out, I'm interested to hear your results!
